_posts/2015-08-25-building-on-the-bones-of-your-drafts.markdown

```diff
@@ -0,0 +1,77 @@
+---
+layout: post
+title: "Building on the bones of your drafts"
+date: 2015-08-25T20:06:47-05:00
+---
+
+In my [last post](http://designisrefactoring.com/2015/08/03/the-power-of-drafts/) I talked about how our rough drafts have the power to reveal our humanity and our learning processes. But I never talked about how we actually _improve_ our code, how we can build on our flawed early implementations and make them better.
+
+Just as a writer doesn't randomly change words and call the result a Second Draft, we can't improve our code by just randomly tweaking things and pushing it to GitHub. Your commit log can be endlessly long and the code might still be a First Draft. So, how do we actually move from First Draft to Second? How do we build upon the bones of our drafts, each new layer better than the ones beneath?
+
+<!--break-->
+
+I quickly mentioned a process in my last post, but didn't discuss it in any depth. Here it is again:
+
+1. Start with a vague idea
+2. Write some awful, ugly code
+3. Use the knowledge I gained in step two to write some slightly better code
+4. Repeat...forever
+
+Step 3 could easily be renamed "Refactor!" but I've found that word can be misunderstood. For this process to work -- for your code to actually improve --  refactoring has to be done correctly. Your knowledge of the system has to change, and that new knowledge has to be applied to your code.
+
+Let's look at a silly example:
+
+I have an app that lets people manage their lemur collections. Lemurs (or their owners) get bored of names pretty quickly, so they want to be able to rename all of their lemurs at once using randomly picked names. We follow the normal TDD workflow, writing some tests and then some code:
+
+```
+def lemur_renamer(lemurs)
+  lemurs.each do |x|
+    x.update_name(LemurNames.a_lemur_name)
+  end
+end
+```
+
+We've done Red and Green, now it's time to Refactor! With our current knowledge of the system, what can we change here?
+
+```
+def lemur_renamer(lemurs)
+  lemurs.each { |lemur| lemur.update_name(LemurNames.random) }
+end
+```
+
+We've renamed the useless variable `x` to the descriptive `lemur` and renamed the incorrect `a_lemur_name` to the more accurate `random`. And we altered the code to use a more idiomatic block instead of `do/end`.
+
+All of these changes are good. But none of them are refactoring. No new knowledge of how the code should work was introduced, and no aspect of the code's design was changed.
```

Comment:
Hi @IanWhitney! I enjoyed reading your latest posts, but I couldn't agree to this sentence. I agree with the general idea you are trying to convey in the post. However, saying that those changes are not refactoring, is not correct in my opinion.

Looking at the definition of [refactoring](http://martinfowler.com/bliki/DefinitionOfRefactoring.html) by Martin Fowler, you could say that those changes "make it easier to understand" and that the changes below "make it (...) cheaper to modify without changing its observable behavior".

I understand that your method is just an example which is very small but in a real application a method can get long and hard to understand rather quickly. Here is a couple of examples that wouldn't fit your definition of refactoring:

* http://www.martinfowler.com/articles/refactoring-pipelines.html
* http://www.martinfowler.com/articles/refactoring-external-service.html

No new knowledge was gained in any of those examples (if I remember correctly).

Refactoring is one of those words that a lot of people misunderstand and this post could be very confusing to people new to software development. That is why we should try to keep our definitions correct so that we can communicate easily.

My suggestion would be to either: rewrite the post to remove the incorrect references to refactoring; or add a warning at the beginning of the post stating that this is **your** definition of refactoring and not the standard definition of refactoring.